**1. Read data from CSV file using pandas**

In [1]:
from google.colab import drive
drive.mount("/drive", force_remount=True)
import csv
import pandas as pd
fakeArticle = pd.read_csv('/drive/My Drive/Colab Notebooks/IST 664/2023/Fake.csv')

#Extract onlt the text field from the data"
fake = fakeArticle["text"]
print(fake)
len(fake), type(fake[0])

ModuleNotFoundError: No module named 'google'

**2.PRE-PROCESSING**

Merge the data from all articles to one single string

In [2]:
fakeText = ''
for i in range(len(fake)):
    fakeText += fake[i]

print(fakeText[0:1000])
len(fakeText)

Donald Trump just couldn t wish all Americans a Happy New Year and leave it at that. Instead, he had to give a shout out to his enemies, haters and  the very dishonest fake news media.  The former reality show star had just one job to do and he couldn t do it. As our Country rapidly grows stronger and smarter, I want to wish all of my friends, supporters, enemies, haters, and even the very dishonest Fake News Media, a Happy and Healthy New Year,  President Angry Pants tweeted.  2018 will be a great year for America! As our Country rapidly grows stronger and smarter, I want to wish all of my friends, supporters, enemies, haters, and even the very dishonest Fake News Media, a Happy and Healthy New Year. 2018 will be a great year for America!  Donald J. Trump (@realDonaldTrump) December 31, 2017Trump s tweet went down about as welll as you d expect.What kind of president sends a New Year s greeting like this despicable, petty, infantile gibberish? Only Trump! His lack of decency won t eve

59815411

There are 59815411 characters in all the articles together

Tokenization

To separate data into individual words I am using "punkt" tokenizer which separates by spaces and special characters(non-alphabetical characters)

In [4]:
import nltk
nltk.download('punkt')
fakeTokens=nltk.word_tokenize(fakeText)
print(len(fakeTokens), fakeTokens[0:1000])

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


11020065 ['Donald', 'Trump', 'just', 'couldn', 't', 'wish', 'all', 'Americans', 'a', 'Happy', 'New', 'Year', 'and', 'leave', 'it', 'at', 'that', '.', 'Instead', ',', 'he', 'had', 'to', 'give', 'a', 'shout', 'out', 'to', 'his', 'enemies', ',', 'haters', 'and', 'the', 'very', 'dishonest', 'fake', 'news', 'media', '.', 'The', 'former', 'reality', 'show', 'star', 'had', 'just', 'one', 'job', 'to', 'do', 'and', 'he', 'couldn', 't', 'do', 'it', '.', 'As', 'our', 'Country', 'rapidly', 'grows', 'stronger', 'and', 'smarter', ',', 'I', 'want', 'to', 'wish', 'all', 'of', 'my', 'friends', ',', 'supporters', ',', 'enemies', ',', 'haters', ',', 'and', 'even', 'the', 'very', 'dishonest', 'Fake', 'News', 'Media', ',', 'a', 'Happy', 'and', 'Healthy', 'New', 'Year', ',', 'President', 'Angry', 'Pants', 'tweeted', '.', '2018', 'will', 'be', 'a', 'great', 'year', 'for', 'America', '!', 'As', 'our', 'Country', 'rapidly', 'grows', 'stronger', 'and', 'smarter', ',', 'I', 'want', 'to', 'wish', 'all', 'of', 'my

Lower-case

To maintain uniformity and consistency among the tokens format for better understanding and processing I am using islower() which makes all token to lower case leters.

In [5]:
fakeWords = [w.lower() for w in fakeTokens]
print(fakeWords[0:1000])

['donald', 'trump', 'just', 'couldn', 't', 'wish', 'all', 'americans', 'a', 'happy', 'new', 'year', 'and', 'leave', 'it', 'at', 'that', '.', 'instead', ',', 'he', 'had', 'to', 'give', 'a', 'shout', 'out', 'to', 'his', 'enemies', ',', 'haters', 'and', 'the', 'very', 'dishonest', 'fake', 'news', 'media', '.', 'the', 'former', 'reality', 'show', 'star', 'had', 'just', 'one', 'job', 'to', 'do', 'and', 'he', 'couldn', 't', 'do', 'it', '.', 'as', 'our', 'country', 'rapidly', 'grows', 'stronger', 'and', 'smarter', ',', 'i', 'want', 'to', 'wish', 'all', 'of', 'my', 'friends', ',', 'supporters', ',', 'enemies', ',', 'haters', ',', 'and', 'even', 'the', 'very', 'dishonest', 'fake', 'news', 'media', ',', 'a', 'happy', 'and', 'healthy', 'new', 'year', ',', 'president', 'angry', 'pants', 'tweeted', '.', '2018', 'will', 'be', 'a', 'great', 'year', 'for', 'america', '!', 'as', 'our', 'country', 'rapidly', 'grows', 'stronger', 'and', 'smarter', ',', 'i', 'want', 'to', 'wish', 'all', 'of', 'my', 'frien

Removing non-words(non-alphabetical words) or Getting Total words
as non-alphabetical characters don't add any weightage to the content

In [6]:
resivedfakeWords = [w for w in fakeWords if w.isalpha()]
print(len(resivedfakeWords))

9579621


There are 9579621 together in all the articles together

Removing stop-words


Removing words which doesn't add any weightage to the content of the sentence

In [7]:
nltk.download('stopwords')
nltk_stops = nltk.corpus.stopwords.words('english')
fakecontentw = [w for w in resivedfakeWords if w not in nltk_stops] #content words
print(len(fakecontentw))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


5168292


There are 5168292 words after removing stop words in the articles

Lemmatization to reduce words to their root word("lemma") despite spellings and pronounciations

In [8]:
nltk.download('wordnet')
nltk.download('omw-1.4')
wnl = nltk.WordNetLemmatizer()
type(wnl)

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


nltk.stem.wordnet.WordNetLemmatizer

In [9]:
fakelemma =[wnl.lemmatize(t) for t in fakeTokens]

#percentage reduction achieved with lemmatization
round( 100 - (len(set(fakelemma))/len(set(fakeTokens)) * 100), 1)

2.9

**3. DATA ANALYSIS**

List of top50 stop words and content words by their frequencies

In [11]:
from nltk import FreqDist
fakestopw = [w for w in fakeWords if w in nltk_stops] #stop words

fakestopwfdist = FreqDist(fakestopw)
fakecontentwfdist = FreqDist(fakecontentw)

topfakestopw = fakestopwfdist.most_common(50)
topfakecontentw = fakecontentwfdist.most_common(50)

print("top 50 stop words")
for pair in topfakestopw:
    print (pair)

print("top 50 content words")
for pair in topfakecontentw:
    print (pair)

top 50 stop words
('the', 527326)
('to', 289638)
('of', 235777)
('and', 224519)
('a', 210576)
('in', 164579)
('that', 150256)
('s', 129631)
('is', 110825)
('for', 92017)
('on', 81064)
('it', 78808)
('he', 77607)
('was', 67786)
('with', 62876)
('his', 58036)
('as', 55766)
('this', 55177)
('be', 48557)
('by', 47334)
('not', 46802)
('are', 46336)
('have', 46027)
('they', 45176)
('i', 43726)
('who', 42270)
('has', 42252)
('at', 41516)
('from', 40740)
('t', 40531)
('you', 40191)
('we', 38002)
('an', 34808)
('about', 32826)
('but', 31403)
('their', 30202)
('she', 25745)
('her', 25701)
('or', 24671)
('what', 24475)
('all', 24430)
('will', 24044)
('been', 22991)
('out', 22917)
('more', 22371)
('if', 21881)
('were', 21607)
('when', 21044)
('can', 20690)
('had', 20425)
top 50 content words
('trump', 74038)
('said', 31125)
('people', 25997)
('president', 25591)
('would', 23457)
('one', 22881)
('clinton', 18074)
('obama', 17865)
('like', 17643)
('donald', 16188)
('also', 15243)
('new', 14173)
('us

From the results, it's clear that after removing stop words and non-alphabetical characters, the frequency ditribution of 50 most common words is the words that hold meaning to the sentence/corpus and also the frequency distribution of stop words(verbose) is more than frequery distribution of content words

List of top50 bigrams by frequencies

In [16]:
import re
from nltk.collocations import *
from nltk.collocations import BigramAssocMeasures
from nltk.collocations import BigramCollocationFinder

def alpha_filter(w):
  # pattern to match word of non-alphabetical characters
  pattern = re.compile('^[^a-z]+$')
  if (pattern.match(w)):
    return True
  else:
    return False

bigram_measures = nltk.collocations.BigramAssocMeasures()
finder = BigramCollocationFinder.from_words(fakeWords)
scored1 = finder.score_ngrams(bigram_measures.raw_freq)
finder.apply_word_filter(alpha_filter) #filter non-alphabetical characters
scored2 = finder.score_ngrams(bigram_measures.raw_freq)
finder.apply_word_filter(lambda w: w in nltk_stops) #filter stop words
scored3 = finder.score_ngrams(bigram_measures.raw_freq)

print("Top 50 Bigrams by frequency")
for i in range(50):
    print (scored1[i])

print("Top 50 Bigrams by frequemcy after removing punctuations")
for i in range(50):
    print (scored2[i])

print("Top 50 Bigrams by frequency after removing punctuations and stop words")
for i in range(50):
    print (scored3[i])

Top 50 Bigrams by frequency
(('of', 'the'), 0.004887629973144442)
(('in', 'the'), 0.003538817602255522)
((',', 'and'), 0.0034519760092159165)
((',', 'the'), 0.002693178307024505)
(('to', 'the'), 0.0025075169701812105)
(('.', 'the'), 0.0023100589697066217)
(('on', 'the'), 0.001688828514169381)
(('(', '@'), 0.0016125131748315458)
(('it', 's'), 0.001543548064371671)
(('to', 'be'), 0.001488013001738193)
((',', 'but'), 0.0014498099602860781)
(('for', 'the'), 0.0014387392451859404)
(('that', 'the'), 0.0013070703303474163)
(('and', 'the'), 0.0012742211593125812)
(('trump', 's'), 0.00120552827955189)
(('.', 'he'), 0.0011901926168312075)
(('.', 'it'), 0.001177125543270389)
(('.', 'i'), 0.001163967726143176)
(('at', 'the'), 0.001148269089156915)
(('donald', 'trump'), 0.0011043492030219423)
(('with', 'the'), 0.0010651479823394871)
((',', 'he'), 0.0010394675530498233)
(('in', 'a'), 0.0010280338636841071)
(('is', 'a'), 0.0009880159509041008)
((',', 'a'), 0.0009442775519019171)
(('from', 'the'), 0.0

From result, we can say that frequency distribution of bigram with punctuations and stops words is more than frequery distribution of bigram after removing

Top 50 bigrams by PMI measure

In [18]:
bigram_measures = nltk.collocations.BigramAssocMeasures()
finder4 = BigramCollocationFinder.from_words(fakeWords)
finder4.apply_freq_filter(5)
fakepmi = finder4.score_ngrams(bigram_measures.pmi)

print("Top 50 PMI Measure")
for i in range(50):
    print (fakepmi[i])

Top 50 PMI Measure
(('//t.co/ltdtbehhgh', 'pic.twitter.com/t2s8ufif5o'), 21.071701302752402)
(('0000', '0907'), 21.071701302752402)
(('0907', '84b4'), 21.071701302752402)
(('314a', '3453'), 21.071701302752402)
(('3453', '0000'), 21.071701302752402)
(('4d6c', '6330'), 21.071701302752402)
(('5707', '5736'), 21.071701302752402)
(('5774', '6a7a'), 21.071701302752402)
(('6330', '666b'), 21.071701302752402)
(('666b', '314a'), 21.071701302752402)
(('6a7a', '4d6c'), 21.071701302752402)
(('7616', '86f7'), 21.071701302752402)
(('84b4', 'f787'), 21.071701302752402)
(('86f7', 'a737'), 21.071701302752402)
(('a737', '5707'), 21.071701302752402)
(('acab', 'pic.twitter.com/naqnehnd5g'), 21.071701302752402)
(('f787', '7616'), 21.071701302752402)
(('kambree', 'kawahine'), 21.071701302752402)
(('kawahine', 'koa'), 21.071701302752402)
(('lynnette', 'hardway'), 21.071701302752402)
(('managementvideo', 'solutionsvideo'), 21.071701302752402)
(('myocardial', 'infarction'), 21.071701302752402)
(('palos', 'verd

Bigrams from PMI are different from bigrams by frequency and bigrams by PMI have very high PMI score(in the range of 20)

Top 50 Adjective words

I am getting the top50 adjective words by using pos_tag

In [20]:
from nltk import pos_tag
nltk.download('averaged_perceptron_tagger')
fake_postags = pos_tag(fakeTokens)
fakeadjtags = [ w for w in fake_postags if w[1] == 'JJ']
fakeadjdist = FreqDist(fakeadjtags)
faketopadjdist = fakeadjdist.most_common(50)
for i in faketopadjdist:
  print(i)

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


(('s', 'JJ'), 19266)
(('other', 'JJ'), 12822)
(('t', 'JJ'), 12806)
(('many', 'JJ'), 8904)
(('American', 'JJ'), 7829)
(('political', 'JJ'), 7288)
(('new', 'JJ'), 7277)
(('last', 'JJ'), 7109)
(('former', 'JJ'), 6494)
(('first', 'JJ'), 6285)
(('own', 'JJ'), 5967)
(('same', 'JJ'), 5295)
(('white', 'JJ'), 5198)
(('presidential', 'JJ'), 4948)
(('black', 'JJ'), 4884)
(('good', 'JJ'), 4848)
(('public', 'JJ'), 4625)
(('federal', 'JJ'), 4620)
(('such', 'JJ'), 4468)
(('much', 'JJ'), 4243)
(('Russian', 'JJ'), 3752)
(('national', 'JJ'), 3675)
(('Democratic', 'JJ'), 3549)
(('real', 'JJ'), 3396)
(('great', 'JJ'), 3332)
(('few', 'JJ'), 3190)
(('clear', 'JJ'), 3167)
(('military', 'JJ'), 3132)
(('little', 'JJ'), 3078)
(('social', 'JJ'), 3022)
(('next', 'JJ'), 2986)
(('big', 'JJ'), 2947)
(('several', 'JJ'), 2879)
(('illegal', 'JJ'), 2876)
(('recent', 'JJ'), 2798)
(('foreign', 'JJ'), 2761)
(('conservative', 'JJ'), 2716)
(('only', 'JJ'), 2642)
(('able', 'JJ'), 2602)
(('sure', 'JJ'), 2601)
(('entire', 'JJ')

**Counts**

In [21]:
fakeArticleTokens = [None]*len(fake)
for i in range(len(fake)):
    fakeArticleTokens[i]=nltk.word_tokenize(fake[i]);
print(len(fakeArticleTokens))
print(fakeArticleTokens[0],len(fakeArticleTokens[0]))

23481
['Donald', 'Trump', 'just', 'couldn', 't', 'wish', 'all', 'Americans', 'a', 'Happy', 'New', 'Year', 'and', 'leave', 'it', 'at', 'that', '.', 'Instead', ',', 'he', 'had', 'to', 'give', 'a', 'shout', 'out', 'to', 'his', 'enemies', ',', 'haters', 'and', 'the', 'very', 'dishonest', 'fake', 'news', 'media', '.', 'The', 'former', 'reality', 'show', 'star', 'had', 'just', 'one', 'job', 'to', 'do', 'and', 'he', 'couldn', 't', 'do', 'it', '.', 'As', 'our', 'Country', 'rapidly', 'grows', 'stronger', 'and', 'smarter', ',', 'I', 'want', 'to', 'wish', 'all', 'of', 'my', 'friends', ',', 'supporters', ',', 'enemies', ',', 'haters', ',', 'and', 'even', 'the', 'very', 'dishonest', 'Fake', 'News', 'Media', ',', 'a', 'Happy', 'and', 'Healthy', 'New', 'Year', ',', 'President', 'Angry', 'Pants', 'tweeted', '.', '2018', 'will', 'be', 'a', 'great', 'year', 'for', 'America', '!', 'As', 'our', 'Country', 'rapidly', 'grows', 'stronger', 'and', 'smarter', ',', 'I', 'want', 'to', 'wish', 'all', 'of', 'my', 

In [35]:
import re
def alpha_filter(w):
  # pattern to match word of non-alphabetical characters
  pattern = re.compile('^[^a-z]+$')
  if (pattern.match(w)):
    return True
  else:
    return False

def punctuations(w):
  # pattern to match word of non-alphabetical characters
  pattern = re.compile('^[^a-z]+$')
  if (not(w.isnumeric()) and pattern.match(w)):
    return True
  else:
    return False

def caps_letter(w):
  if (w !="I" and w != 'i'):
      if(w.isupper()):
          return True
      else:
          return False
  else:
    return False


def exclamation(w):
    if( '!' in w):
        return True
    else:
        return False

faketotalwords = []*len(fakeArticleTokens)
fakecontentwords = [] * len(fakeArticleTokens)
faketotalcaps  = [] * len(fakeArticleTokens)
faketotalexclaim  = [] * len(fakeArticleTokens)
faketotalpunct = [] * len(fakeArticleTokens)
for i in range(len(fakeArticleTokens)):
    currfaketotalwords = 0
    currfakecontentwords = 0
    currfakecapswords = 0
    currfakeexclaimwords = 0
    currfakepunctwords = 0
    for j in  range(len(fakeArticleTokens[i])):
        w = fakeArticleTokens[i][j]
        if (not(punctuations(w))):
            currfaketotalwords+=1
        if (not(punctuations(w)) and w not in nltk_stops):
            currfakecontentwords+=1
        if (caps_letter(w)):
            currfakecapswords +=1
        if (exclamation(w)):
            currfakeexclaimwords +=1
        if( punctuations(w)):
            currfakepunctwords +=1
    faketotalwords.append(currfaketotalwords)
    fakecontentwords.append(currfakecontentwords)
    faketotalcaps.append(currfakecapswords)
    faketotalexclaim.append(currfakeexclaimwords)
    faketotalpunct.append(currfakepunctwords)

In [36]:
print("Total number of words")
for i in range(len(faketotalwords)):
    print (faketotalwords[i])

print("Total number of content words")
for i in range(len(fakecontentwords)):
    print (fakecontentwords[i])

print("Total number of capitalized words")
for i in range(len(faketotalcaps)):
    print (faketotalcaps[i])

print("Total number of exclaimation words")
for i in range(len(faketotalexclaim)):
    print (faketotalexclaim[i])

print("Total number of punctuations")
for i in range(len(faketotalpunct)):
    print (faketotalpunct[i])

Streaming output truncated to the last 5000 lines.
156
44
5
46
42
39
40
41
135
28
45
141
110
105
31
61
72
43
65
27
4
38
50
22
29
47
62
55
36
63
35
19
14
71
72
46
36
164
31
57
8
83
13
22
53
30
50
38
60
17
49
26
23
45
38
34
46
36
128
18
159
27
74
16
52
29
20
12
22
47
74
46
48
111
26
148
35
79
41
102
38
58
32
55
40
25
17
44
84
32
45
16
38
22
31
25
32
13
5
29
69
44
153
99
21
11
181
85
37
94
41
65
21
158
47
29
36
43
26
33
117
60
41
59
14
43
51
29
8
79
50
26
185
18
4
33
13
261
431
41
12
198
35
29
55
13
89
26
54
31
52
46
61
10
59
51
36
31
12
68
42
14
32
43
38
94
57
81
37
212
247
51
0
212
52
12
14
28
80
74
32
24
26
30
20
51
83
29
30
36
9
31
31
49
138
53
33
105
32
29
35
13
148
30
10
33
30
41
68
21
69
22
70
63
19
105
60
87
20
45
88
21
79
24
5
101
56
27
20
56
91
64
27
95
38
31
105
33
19
66
34
28
25
26
35
28
10
61
132
22
15
71
7
58
35
41
21
29
73
37
58
9
18
74
8
2
0
72
51
24
71
45
24
66
13
176
7
17
22
75
41
214
48
167
71
85
75
33
53
51
129
32
32
79
131
80
85
116
28
119
25
51
44
196
18
39
27
113
20

Saving the counts to CSV file

In [37]:
csv_file = '/drive/My Drive/Colab Notebooks/IST 664/2023/Fake.csv'
column_name1 = "TotalWords"
column_name2 = "TotalcontentWords"
column_name3 = "TotalcapitalizedWords"
column_name4 = "TotalexclaimationWords"
column_name5 = "TotalPunctuations"
fakeArticle[column_name1] = faketotalwords
fakeArticle[column_name2] = fakecontentwords
fakeArticle[column_name3] = faketotalcaps
fakeArticle[column_name4] = faketotalexclaim
fakeArticle[column_name5] = faketotalpunct

# Save the DataFrame to the CSV file
fakeArticle.to_csv(csv_file, index=False)

Please find the data interpretation in true.ipynb file